src/commands/publish/publish.command.ts

```diff
@@ -0,0 +1,39 @@
+import * as _path from 'path';
+import { ICommand } from '../command';
+import { NodePublishCommand } from './node/node-publish.command';
+import { DotnetPublishCommand } from './dotnet/dotnet-publish.command';
+import { IFileService, FileService } from '../../services/file.service';
+
+export class PublishCommand implements ICommand {
+
+    get name(): string { return "publish" };
+    publishCommands: ICommand[] = [
+        new NodePublishCommand(),
+        new DotnetPublishCommand()
+    ];
+    fileService: IFileService = new FileService();
+
+    run = (environment: string = "*", solutionFilePath: string = "./shaman.json"): Promise<void> => {
+        console.log("Publishing solution...");
+        let publishEnvironmentsTask = this.fileService.getShamanFile(solutionFilePath).then(solution => {
+            if (environment != "*") return this.publishEnvironment(environment, solutionFilePath);
+            let projectEnvironments = solution.projects.map(p => p.environment);
+            let environmetsSet = [...new Set(projectEnvironments)];
+            return environmetsSet.reduce((a, b) => a.then(_ =>
+                this.publishEnvironment(b, solutionFilePath)
+            ), Promise.resolve());
+        });
+        return publishEnvironmentsTask.then(_ => console.log("Solution publish is complete."));
+    }
+
+    private publishEnvironment(environment: string, solutionFilePath: string): Promise<void> {
+        let cmd = this.publishCommands.find(c => c.name == `publish-${environment}`)
+        if (!cmd) return Promise.reject(new Error(`Invalid environment '${environment}'.`));
+        let cwd = solutionFilePath.replace("shaman.json", "");
+        return this.fileService.pathExists(`${cwd}/bin`)
+            .then(exists => { if (!exists) this.fileService.createFolder(cwd, "bin") })
```

Comment:
It looks like this will be called once for every environment, but only needs to be called once for the whole solution. This can also be made easier by exporting an "ensureFolderExists" method on the file service (more details in other comment)